### PR TITLE
preserve ownership/permissions on backup and restore

### PR DESCRIPTION
**Pull request by MaximilianMeister**
cc @vuntz

### CODE REVIEW

**Comment by vuntz** [comment]
`system("sudo", "chown", "-R", "crowbar:crowbar", dest)` in export.rb will need to be removed.

File: crowbar_framework/app/models/backup.rb [hunk 1]
```diff
@@ -100,13 +100,21 @@ def create_archive
 
     Crowbar::Backup::Export.new(dir).export
     Dir.chdir(dir) do
-      ar = Archive::Compress.new(
+      cmd = [
+        "tar",
+        "--same-permissions",
+        "--same-owner",
```
Comment (vuntz) [inline crowbar_framework/app/models/backup.rb]: There's no option in libarchive for that?

**Comment by vuntz** [comment]
Also, when we extract the backup, the files are created with the wrong ownership (maybe correct permissions, though).

File: crowbar_framework/app/models/backup.rb (hunk 1 above)
Comment (MaximilianMeister) [inline crowbar_framework/app/models/backup.rb]: the problem I thought started with trying to pack files that don't belong to crowbar. I assumed libarchive does not have the permissions to pack them, that's why I used a system call (which in return makes me wonder why I haven't used sudo here...).
I guess I was wrong. It seems we can keep the code here as it was, I will have to test it

**Comment by MaximilianMeister** [comment]
> system("sudo", "chown", "-R", "crowbar:crowbar", dest) in export.rb will need to be removed.

right, thanks I totally forgot that one

**Comment by MaximilianMeister** [comment]
> Also, when we extract the backup, the files are created with the wrong ownership (maybe correct permissions, though).

I think the only place where the tar flags matter is the extract case. from the manpage
```
       -p, --preserve-permissions, --same-permissions
              extract information about file permissions (default for superuser)
       --same-owner
              try extracting files with the same ownership as exists in the archive (default for superuser)
```

so I kept the `create_archive` as it is (using libarchive), and just changed the `extract` method to use tar.
`rubygem-archive` does have a flag for permission but not for ownership it seems. https://github.com/erikh/archive/blob/master/lib/archive/extract.rb#L36

**Comment by vuntz** [comment]
Cool. However, no you can't rely on garbage collection to remove the extracted directory, since you need sudo to remove it.

**Comment by MaximilianMeister** [comment]
> Cool. However, no you can't rely on garbage collection to remove the extracted directory, since you need sudo to remove it.

that makes sense, but this is only the case for the `tmpdir` that was created by `extract`, so where would i hook that? the memoization confuses a bit, how can I know that I can clean the `tmpdir`?

**Comment by rsalevsky** [comment]
I don't like this approach, I prefer when we collect the permission while https://github.com/crowbar/crowbar-core/blob/master/crowbar_framework/lib/crowbar/backup/export.rb#L125 in an yaml file and just use `sudo` to reset the permissions in https://github.com/crowbar/crowbar-core/blob/master/crowbar_framework/lib/crowbar/backup/restore.rb#L67. It's less code changes and the  code is also cleaner.

:-1:

**Comment by vuntz** [comment]
@rsalevsky that's what I discussed with @MaximilianMeister some days ago: what you suggest is reinventing a format to backup (and restore) files with all metadata. Do we really want to do that ourselves and make sure we get everything right, while tar does it already and is known to work well?

**Comment by MaximilianMeister** [comment]
> I don't like this approach, I prefer when we collect the permission while https://github.com/crowbar/crowbar-core/blob/master/crowbar_framework/lib/crowbar/backup/export.rb#L125 in an yaml file and just use sudo to reset the permissions in https://github.com/crowbar/crowbar-core/blob/master/crowbar_framework/lib/crowbar/backup/restore.rb#L67. It's less code changes and the code is also cleaner.

can you submit your idea? to me it feels like more work/code than this way. and you should know the model better now as you have refactored it

**Comment by rsalevsky** [comment]
@vuntz @MaximilianMeister Ruby gives us access to the meta data so getting and setting this data should be no problem. I did it already in an other project and there was no problem. But I see the point that we should not take the risk so shortly while the release, I will give my +1 when we make sure that we take a look at this after the release.

**Comment by vuntz** [comment]
+1

**Comment by rhafer** [comment]
This needs a rebase, but you might want to wait with that until #227 is merged.

**Pull request closed without merging**